tuning/tune.py

```diff
@@ -820,6 +825,15 @@ def generate_constraints(
         )
     ]
 
+    # The shared memory constraint is best compatible with mmt.
+    # For other dispatch types, there are cases where the shared memory size
+    # may not align with the tile size.
+    if problem_size.dispatch_kind == DispatchKind.mmt:
+        shared_memory = (m * k * (problem_size.lhs_type.bitwidth / 8)) + (
+            k * n * (problem_size.rhs_type.bitwidth / 8)
+        )
+        constraints += [shared_memory <= 65536]
```

Comment:
We should write some tests for this. We can have a helper function that calculates the shared memory usage based on the problem size and (m, n, k), and pass in real values (int) in tests instead of  `z3.Int`.